Feature/water quality tiles

jtarricone opened the pull request:
This builds on #1471 to add rendered tile overlays to the map.

To test:
* clone this branch and bring up an environment (if you have one running, `tiler` will probably have to be reprovisioned to pick up `underscore`, or else you can just `npm install underscore` on the VM
* navigate to `http://localhost:8000`
* try enabling the new layers from the `Overlay` tab in the layer selector; the new ones are "Deleware River Basin TP Concentration", "Deleware River Basin TN Concentration", & "Deleware River Basin TSS Concentration"; cycle through different zoom levels for each, and try switching base maps at some point
* the stream layers should render and the Delaware basin tiles should have different colors per the data binding for each layer

This will probably have merge conflicts with `develop` once #1499 gets merged in, but I think I replicated most of the parts of `styles.mss` to avoid anything gnarly. Additionally, it's probably worth taking a critical look at the color scheme provided in #1478, since I could understand how the rendered colors might be hard to see on certain base maps at certain zoom levels.

Connects #1479 & #1478

jtarricone commented on src/tiler/styles.mss:
```diff
@@ -44,126 +51,195 @@
    here have appropriate coinciding filters in server.js
 */
+
+#drb_streams_50 {
```
As I mentioned in the comment, most of this file is modified to mirror the changes from #1499, so a lot of these diffs are irrelevant.

kellyi commented on src/mmw/mmw/settings/layer_settings.py:
```diff
@@ -145,6 +145,30 @@
     },
     {
+        'code': 'nhd_quality_tp',
+        'display': 'Delaware River Basin TP Concentration',
```
I believe these should be labeled as for the "National" rather than "Delaware River Basin" data.

kellyi replied:
re: Layer Selector text: this isn't ideal, but we're currently handling the label for the "Delaware River Basin High Resolution Stream Network" by including a `<br />` tag and some spaces as the value for `display` here.

jtarricone replied:
@mmcfarland is it a viable option to quietly show two layers (NHD and then water quality) when the water quality overlay is enabled? It avoids the left join and might still be somewhat fast if the NHD tiles are cached.

mmcfarland replied:
I considered that as well, but that would still leave us in a situation of rendering blue lines for DRB streams that don't have water quality records, which I don't think they want.

kellyi commented:
I'm working on #1482 which is similar to this but for DRB data, so I'm going to rebase my branch on this one.

mmcfarland commented on src/tiler/server.js:
```diff
@@ -63,8 +90,44 @@ var interactivity = {
+                'WHERE stream_order >= ' + stream_order + ') as q';
+        } else if (tableId === 'nhd_quality_tss') {
+            return '(SELECT geom, stream_order, ' + 
+                caseSql(NHD_QUALITY_TP_MAP,'tss_concmgl') + ' AS nhd_qual_grp ' +
```
`NHD_QUALITY_TP_MAP` should be `TSS` I believe

jtarricone replied:
Ack, good catch.

mmcfarland commented on src/tiler/server.js:
```diff
@@ -63,8 +90,44 @@ var interactivity = {
+        }
+
+        if (tableId === 'nhd_quality_tn') {
+            return '(SELECT geom, stream_order, ' + 
```
These three queries look identical save for some parameters, I think they could be DRY'd up so that the query is constructed only once with the relevant parameters passed in.

jtarricone replied:
True, which I did already with `caseSql()` to handle the most annoying part of the query. A case might reasonably be made for using a stored proc here.

arottersman commented on src/tiler/server.js:
```diff
@@ -63,8 +90,44 @@ var interactivity = {
 
-        return '(SELECT geom, stream_order FROM ' + tableName +
-          ' WHERE stream_order >= ' + stream_order + ') as q';
+        var caseSql = function(mapping, field) {
```
Consider renaming this to `waterQualityCaseSql` to specify what it's for/doing

arottersman commented:
Just tested this....
- I wasn't able to get `underscore` to install with `vagrant reload tiler --provision`, but was able to install it by going into the `tiler` VM and running `npm install` in `/vagrant/src/tiler`
- The coloring on the streams looks all right on most basemaps, except the `terrain` one makes the bright yellow pretty difficult to see

<img width="1044" alt="screen shot 2016-10-04 at 6 09 33 pm" src="https://cloud.githubusercontent.com/assets/7633670/19095181/68b2d9c2-8a62-11e6-8c46-f2fa3d857962.png">

- Should this be all red? 

<img width="1038" alt="screen shot 2016-10-04 at 6 08 09 pm" src="https://cloud.githubusercontent.com/assets/7633670/19095179/654b26a4-8a62-11e6-8598-696f5575f0da.png">

- These colors make it look like we bleed google

<img width="1045" alt="screen shot 2016-10-04 at 6 09 10 pm" src="https://cloud.githubusercontent.com/assets/7633670/19095180/671fdd26-8a62-11e6-95b9-f7c160bea3be.png">

hectcastro commented:
`underscore` may not have installed cleanly because the NPM dependencies are shrinkwrapped. There is a process for updating the shrinkwrap manifest:

> 1. Run npm install in the package root to install the current versions of all dependencies.
> 2. Add or update dependencies. npm install --save each new or updated package individually to update the package.json and the shrinkwrap. Note that they must be explicitly named in order to be installed: running npm install with no arguments will merely reproduce the existing shrinkwrap.
> 3. Validate that the package works as expected with the new dependencies.
> 4. Commit the new npm-shrinkwrap.json, and publish your package.

See: https://docs.npmjs.com/cli/shrinkwrap#building-shrinkwrapped-packages

jtarricone commented:
>Should this be all red? 

The TSS and TP color mappings were reversed.

It's probably worth creating an issue to add some kind of legend to the map.

kellyi commented on src/tiler/server.js:
```diff
@@ -154,6 +217,11 @@ var config = {
                 req.params.sql = getSqlForStreamByReq(req);
             }
 
+            if (tableId.indexOf('quality') >= 0) {
```
Consider renaming `quality` to `nhd_quality` or `nhd_quality_t` to narrow down the possible matches for this check.

jtarricone replied:
Ahh good look, I'll update it.

mmcfarland commented on src/tiler/styles.mss:
```diff
@@ -44,126 +51,195 @@
+  line-cap: round;
+}
+
+#nhd_quality_tp {
```
These style declarations look to be repeated since you did the work of coalescing the values into groups in SQL.  You can render all three with one block if you join the selectors:
```css
#nhd_quality_tp, 
#nhd_quality_tn, 
#nhd_quality_tss {
   [nhd_qual_grp="L1"] {
     line-color: @wtrQualColor0;
   }
  ...
}

ajrobbins commented:
@jfrankl Can you weigh in on the colors here? Client requested red-yellow-green ramp, I believe we're using these colors https://github.com/WikiWatershed/model-my-watershed/issues/1478#issuecomment-250470268

mmcfarland commented on src/tiler/server.js:
```diff
@@ -63,8 +90,44 @@ var interactivity = {
+            bounds.push('ELSE \'NA\' ');
+            bounds.push('END');
+
+            return _.reduce(bounds, function(acc, x) { return acc + ' ' + x; });
```
I believe you're reducing this array to join the string elements inside.  You can achieve the same much more simply with `bounds.join(' ')` and we can remove the `underscore` dependency.

mmcfarland commented on src/tiler/styles.mss:
```diff
@@ -36,6 +36,13 @@
+@wtrQualColor2: #ffffbf;
+@wtrQualColor3: #a6d96a;
+@wtrQualColor4: #1a9641;
+@wtrQualColorNA: #648db9;
```
The blue for `NA` here seems a little too in-line with the symbology for actual data and close to the regular stream color.  I'd suggest a dark gray or something to indicate more clearly that there is no water quality data.

jtarricone replied:
I updated it to a greyer grey.

mmcfarland commented:
You'll also want to update the `setupdb.sh` script so that the cached tiles for these layers are purged when we reload data: https://github.com/WikiWatershed/model-my-watershed/blob/develop/scripts/aws/setupdb.sh#L128-L131  (Add the `layerIds` as `PATHS` and add a call to `purge_tile_cache $PATHS`)

jfrankl commented:
Thanks for consulting me about the colors. I will need to know more about this feature before I can think about how it needs to work and how it will fit into the application. I will not have time to provide recommended colors during this sprint, but I can get that for you early this coming sprint, which begins tomorrow.

jtarricone commented:
I refactored to cut underscore out, and while I was changing `setupdb.sh` I noticed that the `PATH`s for the NHD & DRB layers appeared to be wrong, which I would guess would prevent the tile cache from being purged, so I updated those as well.

jtarricone commented:
I think `setupdb.sh` is missing a path for school district boundary tiles to purge, is that intentional?

mmcfarland commented on scripts/aws/setupdb.sh:
```diff
@@ -102,7 +102,7 @@ fi
     # Fetch stream network layer sql files
     FILES=("nhdflowline.sql.gz")
-    PATHS=("stream")
+    PATHS=("nhd_streams_v2")
```
These stream updates already have a commit waiting in #1510

arottersman commented:
These changes look good to me +1

hectcastro commented:
@azavea-bot rebuild

kellyi commented:
Heads up that there's a merge conflict.
